wgsl/index.bs

```diff
@@ -4557,11 +4555,271 @@ TODO: *Stub*: Expression evaluation
+      - Sound (meaning that it rejects every program that would break the uniformity requirements of builtins)
+      - Linear time complexity
+      - Refactoring a piece of code into a function cannot make it invalid if it was valid before
+      - If the analysis refuses a program, it provides a straightforward chain of implications that can be used to craft a good error message
+</div>
+
+The analysis analyzes each function in turn, verifying that there is a context where it is safe to call this function (otherwise it rejects the program), and computing metadata about the function to help analyze its callers in turn. This means that the call graph must first be built, and functions must be analyzed from the leaves (functions that call no function outside the standard library) upwards (toward the entry point), so that whenever we analyze a function we have already computed
+the metadata for all of its callees. There is no risk of being trapped in a cycle, as recurrence is forbidden in the language.
+
+### Analyzing the uniformity requirements of a function ### {#uniformity-function}
+
+Each function is analyzed in two phases.
+
+The first phase walks over the AST of the function, building a directed graph along the way.
+The second phase explores that graph, resulting in either rejecting the program, or computing the constraints on calling this function.
+
+Note: Each vertex in this graph corresponds either to the statement "The control-flow at a specific program point must be uniform" or to "The value of a specific expression must be uniform" or to "A specific variable must always hold a uniform value" or is one of the special vertices True and False.
+      Each (directed) edge corresponds to an implication.
+
+In more detail, here is the algorithm for analyzing a function:
+
+* Create some vertices called "True", "False", "CF_start", "CF_return" and if the function is non-void a vertex called "Value_return"
+* Create one vertex for each parameter of the function which we'll call "arg_i"
+* Walk over the AST of the function, adding vertices and edges to the graph following the rules of the next section.
+* Look at the *Behaviors* of the function body treated as a statement (see next section):
+    * If it has a behavior which is neither Nothing nor Discard, then reject the program
+    * Else remember it for later it.
+* Look at which vertices are reachable from "True"
+    * If this set includes the vertex "False", then reject the program.
+    * Remember for later whether this set includes "CF_start" and/or any of the "arg_i"
+    * (optional, makes the analysis faster) Remove from the graph all vertices that have been visited
+* Look at which vertices are reachable from "CF_return"
+    * Remember for later whether this set includes "False" and/or any of the "arg_i"
+    * (optional, makes the analysis faster) Remove from the graph all vertices that have been visited
+* Look at which vertices are reachable from "Value_return"
```

Comment:
if it exists

Reply:
right, I keep forgetting void functions. Will fix.